opm/simulators/linalg/WellOperators.hpp

```diff
@@ -0,0 +1,239 @@
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_WELLOPERATORS_HEADER_INCLUDED
+#define OPM_WELLOPERATORS_HEADER_INCLUDED
+
+#include <dune/istl/operators.hh>
+
+
+namespace Opm
+{
+
+
+//=====================================================================
+// Implementation for ISTL-matrix based operators
+// Note: the classes WellModelMatrixAdapter and
+// WellModelGhostLastMatrixAdapter were moved from ISTLSolverEbos.hpp
+// and subsequently modified.
+//=====================================================================
+
+// This class is intended to hide the actual type of the well model
+// (which depends on a TypeTag) in the WellModelMatrixAdapter, so that
+// we do not need to instantiate that class (and classes using it) for
+// every TypeTag variant.
+template <class WellModel, class X, class Y>
+class WellModelAsLinearOperator : public Dune::LinearOperator<X, Y>
+{
+public:
+    using Base = Dune::LinearOperator<X, Y>;
+    using field_type = typename Base::field_type;
+
+    explicit WellModelAsLinearOperator(const WellModel& wm)
+        : wellMod_(wm)
+    {
+    }
+    /*! \brief apply operator to x:  \f$ y = A(x) \f$
+          The input vector is consistent and the output must also be
+       consistent on the interior+border partition.
+     */
+    void apply (const X& x, Y& y) const override
+    {
+        wellMod_.apply(x, y );
+    }
+
+    //! apply operator to x, scale and add:  \f$ y = y + \alpha A(x) \f$
+    virtual void applyscaleadd (field_type alpha, const X& x, Y& y) const override
+    {
+        wellMod_.applyScaleAdd( alpha, x, y );
+    }
+
+    /// Category for operator.
+    /// This is somewhat tricky, I consider this operator sequential
+    /// since (unlike WellModelMatrixAdapter) it does not do any
+    /// projections etc. but only forwards the calls to the sequential
+    /// well model.
+    Dune::SolverCategory::Category category() const override
+    {
+        return Dune::SolverCategory::sequential;
+    }
+private:
+    const WellModel& wellMod_;
+};
+
+/*!
+   \brief Adapter to turn a matrix into a linear operator.
+
+   Adapts a matrix to the assembled linear operator interface
+ */
+template<class M, class X, class Y, bool overlapping >
+class WellModelMatrixAdapter : public Dune::AssembledLinearOperator<M,X,Y>
+{
+  typedef Dune::AssembledLinearOperator<M,X,Y> BaseType;
```

Comment:
Remove? It does not seem to be used, but maybe I am missing something.

Reply:
Done